nucliadb/src/nucliadb/common/cache.py

```diff
@@ -0,0 +1,168 @@
+cached_extracted_texts = Gauge("nucliadb_cached_extracted_texts")
+
+
+T = TypeVar("T")
+
+
+class Cache(Generic[T], ABC):
+    """Low-level bounded cache implementation with access to per-key async locks
+    in case cache users want to lock concurrent access.
+
+    This cache is measured using a mandatory metric all subclasses must define.
+
+    """
+
+    def __init__(self, cache_size: int) -> None:
+        self.cache: LRU[str, T] = LRU(cache_size, callback=self._evicted_callback)
+        self.locks: dict[str, asyncio.Lock] = {}
+
+    def get(self, key: str) -> Optional[T]:
+        return self.cache.get(key)
+
+    # Get a lock for a specific key. Locks will be evicted at the same time as
+    # key-value pairs
+    def get_lock(self, key: str) -> asyncio.Lock:
+        return self.locks.setdefault(key, asyncio.Lock())
+
+    def set(self, key: str, value: T):
+        len_before = len(self.cache)
+
+        self.cache[key] = value
+
+        len_after = len(self.cache)
+        if len_after - len_before > 0:
+            self._cache_size_metric.inc(len_after - len_before)
+
+    def contains(self, key: str) -> bool:
+        return key in self.cache
+
+    def clear(self):
+        self._cache_size_metric.dec(len(self.cache))
+        self.cache.clear()
+        self.locks.clear()
+
+    def __del__(self):
+        # we want to clear the cache before deleting the object and set the
+        # metric appropriately
+        self.clear()
+
+    @abstractmethod
+    @cached_property
+    def _cache_size_metric(self) -> Gauge: ...
+
+    def _evicted_callback(self, key: str, value: T):
+        self.locks.pop(key, None)
+
+
+class ResourceCache(Cache[ResourceORM]):
+    # This cache size is an arbitrary number, once we have a metric in place and
+    # we analyze memory consumption, we can adjust it with more knoweldge
+    def __init__(self, cache_size: int = 128) -> None:
+        super().__init__(cache_size)
+
+    @cached_property
+    def _cache_size_metric(self) -> Gauge:
+        return cached_resources
+
+
+class ExtractedTextCache(Cache[ExtractedText]):
+    """
+    Used to cache extracted text from a resource in memory during the process
+    of search results hydration.
+
+    This is needed to avoid fetching the same extracted text multiple times,
+    as matching text blocks are processed in parallel and the extracted text is
+    fetched for each field where the text block is found.
+    """
+
+    def __init__(self, cache_size: int = 128):
+        super().__init__(cache_size)
+
+    @cached_property
+    def _cache_size_metric(self) -> Gauge:
+        return cached_extracted_texts
+
+
+# Global caches (per asyncio task)
+
+rcache: ContextVar[Optional[ResourceCache]] = ContextVar("rcache", default=None)
+etcache: ContextVar[Optional[ExtractedTextCache]] = ContextVar("etcache", default=None)
+
+
+# Cache management
+
+
+# Get or create a resource cache specific to the current asyncio task
```

Comment:
I would add to the comment something like: "... and all its subtasks. Make sure to set the context var in the parent task of the concurrent tasks you want to spawn."